File: conversoes/afn_lambda_para_afn.py
from typing import Dict, Any, Set
from af_utils import fecho_lambda, mover, LAMBDA


def converter(automato: Dict[str, Any]) -> Dict[str, Any]:
    """
    Converte um AFN-λ (AFN com transições lambda) em um AFN equivalente
    sem transições lambda, usando fecho-λ.
    """

    # Remove lambda do alfabeto
    novo_alfabeto = [s for s in automato.get("alfabeto", []) if s != LAMBDA]

    estados = automato.get("estados", [])
    estados_finais_originais = set(automato.get("estados_finais", []))

    # Novo conjunto de transições
    novas_transicoes = []

    # Para cada estado e símbolo, calcula δ'(q, a)
    for estado in estados:
        fecho_q = fecho_lambda(automato, {estado})

        for simbolo in novo_alfabeto:
            destinos = set()

            # move(fecho(q), a)
            move_result = mover(automato, fecho_q, simbolo)

            # fecho_lambda(move(...))
            for p in move_result:
                destinos |= fecho_lambda(automato, {p})

            for destino in destinos:
                novas_transicoes.append([estado, destino, simbolo])

    # Estados iniciais permanecem os mesmos
    novos_estados_iniciais = automato.get("estados_iniciais", [])

    # Novo conjunto de estados finais
    novos_estados_finais = []
    for estado in estados:
        fecho_q = fecho_lambda(automato, {estado})
        if not fecho_q.isdisjoint(estados_finais_originais):
            novos_estados_finais.append(estado)

    novo_automato = {
        "alfabeto": novo_alfabeto,
        "estados": estados,
        "estados_iniciais": novos_estados_iniciais,
        "estados_finais": novos_estados_finais,
        "transicoes": novas_transicoes
    }

    return novo_automato

File: pascal/conversoes/afn_lambda_para_afn.pas
unit afn_lambda_para_afn;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, af_utils;

function Converter(const Automato: TAutomato): TAutomato;

implementation

function Converter(const Automato: TAutomato): TAutomato;
var
  NovoAlfabeto: TStringArray;
  Estados: TStringArray;
  EstadosFinaisOrig: TStringList;
  NovasTransicoes: TTransicoes;
  Estado: string;
  FechoQ, Destinos, MoveResult, FechoP, TempList: TStringList;
  Simbolo: string;
  I, J: Integer;
  NovosFinais: TStringList;
begin
  SetLength(NovoAlfabeto, 0);
  for Estado in Automato.Alfabeto do
    if Estado <> LAMBDA then
    begin
      SetLength(NovoAlfabeto, Length(NovoAlfabeto) + 1);
      NovoAlfabeto[High(NovoAlfabeto)] := Estado;
    end;

  Estados := Automato.Estados;
  EstadosFinaisOrig := StringArrayToList(Automato.EstadosFinais);
  NovosFinais := TStringList.Create;
  NovosFinais.Sorted := True;
  NovosFinais.Duplicates := dupIgnore;

  SetLength(NovasTransicoes, 0);

  for I := 0 to Length(Estados) - 1 do
  begin
    FechoQ := TStringList.Create;
    FechoQ.Sorted := True;
    FechoQ.Duplicates := dupIgnore;
    FechoQ.Add(Estados[I]);
    TempList := FechoLambda(Automato, FechoQ);
    FechoQ.Free;
    FechoQ := TempList;
    try
      for Simbolo in NovoAlfabeto do
      begin
        MoveResult := Mover(Automato, FechoQ, Simbolo);
        Destinos := TStringList.Create;
        Destinos.Sorted := True;
        Destinos.Duplicates := dupIgnore;
        try
          for J := 0 to MoveResult.Count - 1 do
          begin
            FechoP := TStringList.Create;
            FechoP.Sorted := True;
            FechoP.Duplicates := dupIgnore;
            FechoP.Add(MoveResult[J]);
            TempList := FechoLambda(Automato, FechoP);
            FechoP.Free;
            FechoP := TempList;
            try
              Destinos.AddStrings(FechoP);
            finally
              FechoP.Free;
            end;
          end;

          for J := 0 to Destinos.Count - 1 do
          begin
            SetLength(NovasTransicoes, Length(NovasTransicoes) + 1);
            NovasTransicoes[High(NovasTransicoes)].Origem := Estados[I];
            NovasTransicoes[High(NovasTransicoes)].Destino := Destinos[J];
            NovasTransicoes[High(NovasTransicoes)].Simbolo := Simbolo;
          end;
        finally
          Destinos.Free;
          MoveResult.Free;
        end;
      end;

      if EstadosFinaisCompostos(FechoQ, EstadosFinaisOrig) then
        NovosFinais.Add(Estados[I]);
    finally
      FechoQ.Free;
    end;
  end;

  Result.Alfabeto := NovoAlfabeto;
  Result.Estados := Estados;
  Result.EstadosIniciais := Automato.EstadosIniciais;
  Result.EstadosFinais := ListToStringArray(NovosFinais);
  Result.Transicoes := NovasTransicoes;
  NovosFinais.Free;
  EstadosFinaisOrig.Free;
end;

end.
